faculty/datasets/__init__.py

```diff
@@ -190,94 +156,58 @@ def _isfile(project_path, project_id=None, s3_client=None):
         The project to list files from. You need to have access to this project
         for it to work. Defaults to the project set by FACULTY_PROJECT_ID in
         your environment.
-    s3_client : botocore.client.S3, optional
-        Advanced - a specific boto client for AWS S3 to use.
+    object_client : faculty.clients.object.ObjectClient, optional
+        Advanced - can be used to benefit from caching in chain interactions
+        with datasets.
 
     Returns
     -------
     bool
     """
-    if _isdir(project_path, project_id, s3_client):
+    if _isdir(project_path, project_id):
         return False
     matches = ls(
         project_path,
         project_id=project_id,
         show_hidden=True,
-        s3_client=s3_client,
+        object_client=object_client,
     )
     rationalised_path = path.rationalise_projectpath(project_path)
     return any(match == rationalised_path for match in matches)
 
 
-def _create_parent_directories(project_path, project_id, s3_client):
-
-    bucket = _bucket(project_id)
-
-    # Make sure empty objects exist for directories
-    # List once for speed
-    all_objects = set(
-        ls("/", project_id=project_id, show_hidden=True, s3_client=s3_client)
-    )
-
-    for dirname in path.project_parent_directories(project_path):
-
-        if dirname == "/":
-            # Root is not returned by ls
-            continue
-
-        # We're doing this manually instead of using _isdir as _isdir will
-        # return true if '/somedir/myfile' exists, even if '/somedir/' does not
-        if dirname not in all_objects:
-            bucket_path = path.projectpath_to_bucketpath(dirname, project_id)
-            # Directories on S3 are empty objects with trailing '/' on the key
-            s3_client.put_object(Bucket=bucket, Key=bucket_path)
-
-
-def _put_file(local_path, project_path, project_id, s3_client):
-
-    bucket = _bucket(project_id)
-    bucket_path = path.projectpath_to_bucketpath(project_path, project_id)
-
-    if bucket_path.endswith("/"):
-        msg = (
-            "the destination path {} indicates a directory but the "
-            "source path {} is a normal file - please provide a full "
-            "destination path"
-        ).format(repr(project_path), repr(local_path))
-        raise DatasetsError(msg)
-
-    s3_client.upload_file(local_path, bucket, bucket_path)
+def _create_parent_directories(project_path, project_id, object_client):
+    parent_path = path.project_parent_directory(project_path)
```

Comment:
I suggest we just always do:

```python
parent_path = posixpath.dirname(project_path)
object_client.create_directory(project_id, parent_path, parents=True)
```

This should just work under all circumstances, and we can drop `path.project_parent_directory`.